File: web/src/lib/money.ts
import type { Entry, ExchangeRate } from './api/ledger';

export const supportedCurrencies = ['USD', 'EUR', 'CNY', 'CAD'] as const;

// buildRateIndex receives API exchange rates and returns a currency-to-units-per-USD map.
export function buildRateIndex(rates: ExchangeRate[]): Map<string, number> {
  const index = new Map<string, number>();
  for (const rate of rates) {
    const value = Number(rate.unitsPerUsd);
    if (!Number.isFinite(value) || value <= 0) {
      continue;
    }
    index.set(normalizeCurrencyCode(rate.currency), value);
  }

  return index;
}

// compactMoney receives cents and returns a dense center-label amount.
export function compactMoney(cents: number): string {
  return Math.abs(cents / 100).toLocaleString('en-US', { maximumFractionDigits: 0 });
}

// convertEntryAmountCents receives an entry, target currency, and rates and returns converted cents or null when FX is missing.
export function convertEntryAmountCents(entry: Entry, targetCurrency: string, rates: Map<string, number>): number | null {
  const sourceCurrency = normalizeCurrencyCode(entry.transactionCurrency || entry.bookReportingCurrency || targetCurrency);
  return convertCurrencyAmountCents(entry.amountCents, sourceCurrency, targetCurrency, rates, entry.exchangeRate);
}

// convertCurrencyAmountCents receives cents and source/target currencies and returns converted cents or null when FX is missing.
export function convertCurrencyAmountCents(amountCents: number, sourceCurrency: string, targetCurrency: string, rates: Map<string, number>, exchangeRate?: string): number | null {
  const normalizedSource = normalizeCurrencyCode(sourceCurrency);
  const normalizedTarget = normalizeCurrencyCode(targetCurrency);
  if (normalizedSource === normalizedTarget) {
    return amountCents;
  }

  const rate = parseExchangeRate(exchangeRate);
  if (rate?.from === normalizedSource && rate.to === normalizedTarget) {
    return Math.round(amountCents * rate.rate);
  }
  if (rate?.from === normalizedTarget && rate.to === normalizedSource) {
    return Math.round(amountCents / rate.rate);
  }

  const sourceRate = rates.get(normalizedSource);
  const targetRate = rates.get(normalizedTarget);
  if (!sourceRate || !targetRate) {
    return null;
  }

  return Math.round((amountCents * targetRate) / sourceRate);
}

// formatMoney receives cents and an ISO currency code and returns localized currency text.
export function formatMoney(cents: number, currencyCode: string): string {
  return moneyFormatter(currencyCode).format(cents / 100);
}

// normalizeCurrencyCode receives user or API currency text and returns an uppercase code.
export function normalizeCurrencyCode(value: string): string {
  return value.trim().toUpperCase();
}

// parseExchangeRate receives exchange metadata and returns a normalized rate tuple.
function parseExchangeRate(value?: string): { from: string; to: string; rate: number } | null {
  const match = (value ?? '').trim().toUpperCase().match(/^([A-Z]{3})\/([A-Z]{3})=([0-9]+(?:\.[0-9]+)?)$/);
  if (!match) {
    return null;
  }

  const rate = Number(match[3]);
  if (!Number.isFinite(rate) || rate <= 0) {
    return null;
  }

  return {
    from: match[1],
    to: match[2],
    rate,
  };
}

// moneyFormatter receives an ISO currency code and returns a safe localized formatter.
function moneyFormatter(currencyCode: string): Intl.NumberFormat {
  try {
    return new Intl.NumberFormat('en-US', {
      style: 'currency',
      currency: currencyCode,
    });
  } catch {
    return new Intl.NumberFormat('en-US', {
      style: 'currency',
      currency: 'USD',
    });
  }
}

File: backend/internal/ledger/money.go
package ledger

import (
	"math/big"
	"regexp"
	"strings"

	"github.com/Laisky/errors/v2"
)

type exchangeRate struct {
	From string
	To   string
	Rate *big.Rat
}

var exchangeRatePattern = regexp.MustCompile(`^([A-Z]{3})/([A-Z]{3})=([0-9]+(?:\.[0-9]+)?)$`)

var supportedCurrencies = map[string]struct{}{
	"USD": {},
	"EUR": {},
	"CNY": {},
	"CAD": {},
}

// entryAmountInCurrencyCents receives an entry, target currency, and rates and returns converted target minor units.
func entryAmountInCurrencyCents(entry Entry, targetCurrency string, rates map[string]*big.Rat) (int64, error) {
	return convertAmountCents(entry.AmountCents, entry.TransactionCurrency, targetCurrency, entry.ExchangeRate, rates)
}

// convertAmountCents receives an amount, source currency, target currency, exchange metadata, and rates and returns converted cents.
func convertAmountCents(amountCents int64, sourceCurrency string, targetCurrency string, exchangeRateText string, rates map[string]*big.Rat) (int64, error) {
	sourceCurrency = strings.ToUpper(strings.TrimSpace(sourceCurrency))
	targetCurrency = strings.ToUpper(strings.TrimSpace(targetCurrency))
	if !isSupportedCurrency(sourceCurrency) {
		return 0, errors.Wrap(ErrInvalidInput, "source currency is invalid")
	}
	if !isSupportedCurrency(targetCurrency) {
		return 0, errors.Wrap(ErrInvalidInput, "target currency is invalid")
	}
	if sourceCurrency == targetCurrency {
		return amountCents, nil
	}

	amount := big.NewRat(amountCents, 1)
	if strings.TrimSpace(exchangeRateText) != "" {
		rate, err := parseExchangeRate(exchangeRateText)
		if err != nil {
			return 0, err
		}
		switch {
		case rate.From == sourceCurrency && rate.To == targetCurrency:
			amount.Mul(amount, rate.Rate)
		case rate.From == targetCurrency && rate.To == sourceCurrency:
			amount.Quo(amount, rate.Rate)
		default:
			return 0, errors.Wrapf(ErrInvalidInput, "exchange rate %s/%s does not convert %s to %s", rate.From, rate.To, sourceCurrency, targetCurrency)
		}

		return roundRatToInt64(amount), nil
	}

	sourceRate, ok := rates[sourceCurrency]
	if !ok || sourceRate.Sign() <= 0 {
		return 0, errors.Wrapf(ErrInvalidInput, "exchange rate for %s is missing", sourceCurrency)
	}
	targetRate, ok := rates[targetCurrency]
	if !ok || targetRate.Sign() <= 0 {
		return 0, errors.Wrapf(ErrInvalidInput, "exchange rate for %s is missing", targetCurrency)
	}
	amount.Mul(amount, targetRate)
	amount.Quo(amount, sourceRate)

	return roundRatToInt64(amount), nil
}

// parseExchangeRate receives exchange metadata and returns a normalized decimal conversion rate.
func parseExchangeRate(exchangeRateText string) (exchangeRate, error) {
	normalized := strings.ToUpper(strings.TrimSpace(exchangeRateText))
	matches := exchangeRatePattern.FindStringSubmatch(normalized)
	if len(matches) != 4 {
		return exchangeRate{}, errors.Wrap(ErrInvalidInput, "exchange rate must use FROM/TO=rate")
	}

	rate, ok := new(big.Rat).SetString(matches[3])
	if !ok || rate.Sign() <= 0 {
		return exchangeRate{}, errors.Wrap(ErrInvalidInput, "exchange rate must be positive")
	}

	return exchangeRate{
		From: matches[1],
		To:   matches[2],
		Rate: rate,
	}, nil
}

// isSupportedCurrency receives a currency code and reports whether this phase supports it.
func isSupportedCurrency(currency string) bool {
	_, ok := supportedCurrencies[strings.ToUpper(strings.TrimSpace(currency))]
	return ok
}

// roundRatToInt64 receives a rational amount and returns the nearest integer using half-up rounding.
func roundRatToInt64(value *big.Rat) int64 {
	numerator := new(big.Int).Set(value.Num())
	denominator := new(big.Int).Set(value.Denom())
	quotient, remainder := new(big.Int).QuoRem(numerator, denominator, new(big.Int))
	if remainder.Sign() == 0 {
		return quotient.Int64()
	}

	doubleRemainder := new(big.Int).Mul(new(big.Int).Abs(remainder), big.NewInt(2))
	if doubleRemainder.Cmp(denominator) >= 0 {
		if value.Sign() >= 0 {
			quotient.Add(quotient, big.NewInt(1))
		} else {
			quotient.Sub(quotient, big.NewInt(1))
		}
	}

	return quotient.Int64()
}
